firedrake/slate/slac/compiler.py

```diff
@@ -144,82 +215,295 @@ def compile_expression(slate_expr, tsfc_parameters=None):
+                           cpp=True,
+                           include_dirs=inc,
+                           headers=['#include <Eigen/Dense>',
+                                    '#define restrict __restrict'])
+
+    assert len(slate_expr.ufl_domains()) == 1, (
+        "No support for multiple domains yet!"
+    )
+
+    # Send back a "TSFC-like" SplitKernel object with an
+    # index and KernelInfo
+    coeffs = slate_expr.coefficients()
     kinfo = KernelInfo(kernel=op2kernel,
-                       integral_type="cell",
-                       oriented=oriented,
+                       integral_type=builder.integral_type,
+                       oriented=builder.oriented,
                        subdomain_id="otherwise",
                        domain_number=0,
-                       coefficient_map=range(len(slate_expr.coefficients())),
-                       needs_cell_facets=builder.needs_cell_facets)
+                       coefficient_map=tuple(range(len(coeffs))),
+                       needs_cell_facets=builder.needs_cell_facets,
+                       pass_layer_arg=builder.needs_mesh_levels)
+
     idx = tuple([0]*slate_expr.rank)
 
     return (SplitKernel(idx, kinfo),)
 
 
+def extruded_int_horiz_facet(exp, builder, top_sks, bottom_sks,
+                             clist, coordsym, mesh_level_sym):
+    """Generates a code statement for evaluating interior horizontal
+    facet integrals.
+
+    :arg exp: A :class:`TensorBase` expression.
+    :arg builder: A :class:`KernelBuilder` containing the expression context.
+    :arg top_sks: An iterable of index ordered TSFC kernels for the top
+                  kernels.
+    :arg bottom_sks: An iterable of index ordered TSFC kernels for the bottom
+                     kernels.
+    :arg clist: A `list` of coefficient information to pass into the kernel
+                arguments. (This ensures any outside data is passed when
+                appropriate.)
+    :arg coordsym: An `ast.Symbol` object representing coordinate arguments
+                   for the kernel.
+    :arg mesh_level_sym: An `ast.Symbol` representing the mesh level.
+
+    Returns: A COFFEE code statement, updated coefficient info and updated
+             include_dirs
+    """
+    cl = clist
+    t = builder.get_temporary(exp)
+    nlevels = exp.ufl_domain().topological.layers - 1
+
+    incl = []
+    top_calls = []
+    bottom_calls = []
+    for top, bottom in zip(top_sks, bottom_sks):
+        assert top.indices == bottom.indices, (
+            "Top and bottom kernels must have the same indices"
+        )
+        index = top.indices
+
+        for cindex in set(bottom.kinfo.coefficient_map
+                          + top.kinfo.coefficient_map):
+            c = exp.coefficients()[cindex]
+            cl.extend(builder.coefficient(c))
+
+        incl.extend(set(bottom.kinfo.kernel._include_dirs +
+                        top.kinfo.kernel._include_dirs))
```

Comment:
I'm mainly just being careful about getting all the information here. Or as you put it: belt and braces. I will have a look at `gem.utils` for the `OrderedSet` you mentioned above.